### DISCUSSION TITLE
which-key implementation

### DISCUSSION THREAD

**Discussion by kanopo** [nvim-neorg/neorg#773] [Ideas]
Hi,
I found out that with the which-key plugin by folke, the user can see the many key combo related to neorg(for example about todo items ecc).

The problem is that the names showed in the which-key interface are the full name of the command like "core.norg.qol.todo_item....", so basically the user can't even see the final action the keybind takes.

I would like to make a short name for all the key combo related to neorg, if you think that this could be done, i would liek to help.

Thank you
Have a nice day

**Comment by max397574** [comment]
just use the `register` function from which-key?

**Reply by kevclark** [reply]
I started using Neorg recently, and love it, and I am also using which-key, and the full commands from Neorg's key bindings were getting automatically picked up by which-key.  Here's what I did to clean that up.

After looking at https://github.com/nvim-neorg/neorg/wiki/User-Keybinds, it wasn't clear if `function(keybinds)` allowed for a opts table to be passed in. Looking at the code in `neorg/modules/core/keybinds/module.lua` you can see only the map*() functions take in the opts parameter.  So you can make use of any of them to pass in `opts = { desc = "Friendly description" }`.  If would be nice if the remap*() functions also allowed the opts parameter however.

I am using LazyVim, so here is what I did for which-key `register`.

```
  {
    "folke/which-key.nvim",
    opts = function()
      local wk = require("which-key")
      wk.register({
        ["<leader>n"] = { name = "+Neorg" },
        ["gt"] = { name = "+Neorg Tasks" },
      })
    end,
  },
```
And here is the specific Neorg remaps.
```
        ["core.keybinds"] = {
          config = {
            -- not remapping existing keybinds, only adding opts desc for which-key
            hook = function(keybinds)
             keybinds.unmap("norg", "n", "gtd")
             keybinds.map_event("norg", "n", "gtd", "core.norg.qol.todo_items.todo.task_done", { desc = "Task Done" })

             keybinds.unmap("norg", "n", "gtu")
             keybinds.map_event("norg", "n", "gtu", "core.norg.qol.todo_items.todo.task_undone", { desc = "Task Undone" })

             keybinds.unmap("norg", "n", "gtp")
             keybinds.map_event("norg", "n", "gtp", "core.norg.qol.todo_items.todo.task_pending", { desc = "Task Pending" })

             keybinds.unmap("norg", "n", "gth")
             keybinds.map_event("norg", "n", "gth", "core.norg.qol.todo_items.todo.task_on_hold", { desc = "Task On Hold" })

             keybinds.unmap("norg", "n", "gtc")
             keybinds.map_event("norg", "n", "gtc", "core.norg.qol.todo_items.todo.task_cancelled", { desc = "Task Cancelled" })

             keybinds.unmap("norg", "n", "gtr")
             keybinds.map_event("norg", "n", "gtr", "core.norg.qol.todo_items.todo.task_recurring", { desc = "Task Recurring" })

             keybinds.unmap("norg", "n", "gti")
             keybinds.map_event("norg", "n", "gti", "core.norg.qol.todo_items.todo.task_important", { desc = "Task Important" })

             keybinds.unmap("norg", "n", "<leader>nn")
             keybinds.map_event("norg", "n", "<leader>nn", "core.norg.dirman.new.note", { desc = "New Note" })
            end
          },
        },
```

**Reply by max397574** [reply]
you should put that into `config` and not opts @kevclark otherwise yes
this it what I meant

**Reply by bojohnson5** [reply]
Thanks @kevclark this is what I was looking for. I just had to change "core.norg" to "core" for the commands, but otherwise worked like a charm.
